File: leetcode/easy/136-singleNumber.java
// first pass solution
class Solution {
    public int singleNumber(int[] nums) {
        HashMap<Integer, Integer> hashMap = new HashMap<>();
        
        for (int num : nums) {
            if (hashMap.containsKey(num)) {
                int count = hashMap.get(num);
                hashMap.put(num, count+1);
            } else {
                hashMap.put(num, 1);
            }
        }
        
        for (int key : hashMap.keySet()) {
            if (hashMap.get(key) == 1) {
                return key;
            }
        }
        return -1;
    }
}

// second pass solution
class Solution {
    public int singleNumber(int[] nums) {
        int missingDuplicate = 0;
        
        for (int num : nums) {
            missingDuplicate ^= num;
        }
        return missingDuplicate;
    }
}

File: leetcode/easy/136-singleNumber.js
// Given a non-empty array of integers nums, every element appears twice except for one. Find that single one.

// Follow up: Could you implement a solution with a linear runtime complexity and without using extra memory?

// Example 1:

// Input: nums = [2,2,1]
// Output: 1
// Example 2:

// Input: nums = [4,1,2,1,2]
// Output: 4
// Example 3:

// Input: nums = [1]
// Output: 1

// first pass solution
var singleNumber = function(nums) {
    let seen = new Set();
    
    for (let num of nums) {
        if (!seen.has(num)) {
            seen.add(num);
        } else {
            seen.delete(num);
        }
    }
    return Array.from(seen)[0];
};

// second pass solution
var singleNumber = function(nums) {
    let dict = {};
    
    for (let num of nums) {
        if (!(num in dict)) {
            dict[num] = 0;
        }
        dict[num]++;
    }
    
    for (let key in dict) {
        if (dict[key] === 1) {
            return key;
        }
    }
};
